File: resources/gfx/scripts/items/collectibles.lua
local Collectibles = {}

Collectibles.Items = {}

function Collectibles:Spawn(sprite, x, y)
    local item = {
        sprite = sprite,
        x = x,
        baseY = y,
        timer = math.random() * 10,
        bobHeight = 3,
        bobSpeed = 2,
        offsetY = 0
    }

    table.insert(self.Items, item)
end

function Collectibles:Update(dt)
    for _, item in ipairs(self.Items) do
        item.timer = item.timer + dt
        item.offsetY = math.sin(item.timer * item.bobSpeed) * item.bobHeight
    end
end

function Collectibles:Draw()
    for _, item in ipairs(self.Items) do

        -- shadow
        love.graphics.setColor(1, 1, 1, 0.35)
        love.graphics.ellipse(
            "fill",
            item.x,
            item.baseY + 10,
            10,
            4
        )

        -- item
        love.graphics.setColor(1, 1, 1, 1)
        love.graphics.draw(
            item.sprite,
            item.x,
            item.baseY + item.offsetY,
            0,
            1,
            1,
            item.sprite:getWidth() / 2,
            item.sprite:getHeight() / 2
        )
    end
end

return Collectibles

File: src/systems/collectibles.js
export const Collectibles = {
  items: [],

  spawn(sprite, x, y) {
    this.items.push({
      sprite,
      x,
      y,
      baseY: y,
      timer: Math.random() * 10,
      bobHeight: 3,
      bobSpeed: 2,
      offsetY: 0
    });
  },

  update(dt) {
    for (const item of this.items) {
      item.timer += dt;
      item.offsetY =
        Math.sin(item.timer * item.bobSpeed) * item.bobHeight;
    }
  },

  draw(ctx) {
    for (const item of this.items) {

      // shadow (ellipse like your Lua version)
      ctx.fillStyle = "rgba(255, 255, 255, 0.35)";
      ctx.beginPath();
      ctx.ellipse(
        item.x,
        item.baseY + 10,
        10,
        4,
        0,
        0,
        Math.PI * 2
      );
      ctx.fill();

      // item sprite
      if (item.sprite) {
        ctx.drawImage(
          item.sprite,
          item.x - item.sprite.width / 2,
          item.baseY + item.offsetY - item.sprite.height / 2
        );
      }
    }
  }
};